### PR TITLE
NEW Support displaying and setting a default registered method in the member settings fields

### PR DESCRIPTION
**Pull request by robbieaverill**
This PR is split into commits that do various things. A summary:

* Add more strict typing in PHP
* RegisteredMethodManager now assigns the default registered method when registering a method for a member
* Default method is now displayed in the method list in a members profile
* Add "setAsDefault" component and an action to the MFA registered method list (for "Set as default method" action)
* Adjusted removeRegisteredMethod() API endpoint to only accept DELETE requests
* RegisteredMFAMethodListField::getBaseMethods() logic for filteirng the default method was broken because it was expecting an input object, but in practice it was being given a string. Fixing this showed that the logic is no longer required, so I've removed the "filter out default method" part of it
* The "defaultMethod" prop is a string, updated prop types to reflect this, and it is now managed in the Redux store

The default method is set when:

* You set a method as default explicitly
* You remove a method, leaving you with only one method left (except for your backup codes)
* You register your first method

Resolves https://github.com/silverstripe/silverstripe-mfa/issues/66

### CODE REVIEW

File: src/Extension/MemberExtension.php [hunk 1]
```diff
@@ -38,19 +39,35 @@ class MemberExtension extends DataExtension implements PermissionProvider
...
+     *
+     * @param RegisteredMethod $registeredMethod
+     * @return Member
+     * @throws InvalidMethodException
+     */
+    public function setDefaultRegisteredMethod(RegisteredMethod $registeredMethod): Member
+    {
+        if ($registeredMethod->Member()->ID != $this->owner->ID) {
+            throw new InvalidMethodException('The provided method does not belong to this member');
+        }
```
Comment (robbieaverill) [inline src/Extension/MemberExtension.php]: I should write a test for this edge case

File: tests/php/Service/RegisteredMethodManagerTest.php [hunk 2]
```diff
@@ -67,6 +66,27 @@ public function testRegisterForMemberCreatesNewMethod()
...
+        $member->write();
+        $method = new BackupCodeMethod();
+
+        RegisteredMethodManager::singleton()->registerForMember($member, $method, ['foo', 'bar']);
+        $this->assertCount(1, $member->RegisteredMFAMethods());
+        $defaultMethod = $member->getDefaultRegisteredMethod();
+        $this->assertNotNull($defaultMethod, 'Default registered method should have been assigned');
+
+        RegisteredMethodManager::singleton()->registerForMember($member, $method, ['foo', 'baz']);
+        $this->assertCount(1, $member->RegisteredMFAMethods());
```
Comment (robbieaverill) [inline tests/php/Service/RegisteredMethodManagerTest.php]: This should've been 2. It's passing as is which shows it's wrong

File: client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js [hunk 3]
```diff
@@ -268,8 +267,9 @@ class RegisteredMFAMethodListField extends Component {
 
 RegisteredMFAMethodListField.propTypes = {
   backupMethod: registeredMethodShape,
-  defaultMethod: registeredMethodShape,
+  defaultMethod: PropTypes.string,
   readOnly: PropTypes.bool,
+  initialDefaultMethod: PropTypes.string,
   initialRegisteredMethods: PropTypes.arrayOf(registeredMethodShape),
   initialAvailableMethods: PropTypes.arrayOf(availableMethodShape),
   allAvailableMethods: PropTypes.arrayOf(availableMethodShape),
```
Comment (ScopeyNZ) [inline client/src/components/FormField/RegisteredMFAMethodListField/RegisteredMFAMethodListField.js]: Do you want to add the set default endpoint (a couple of lines below)?

File: src/Controller/AdminRegistrationController.php [hunk 4]
```diff
@@ -124,6 +139,13 @@ public function finishRegistration(HTTPRequest $request): HTTPResponse
      */
     public function removeRegisteredMethod(HTTPRequest $request): HTTPResponse
     {
+        if (!$request->isDELETE()) {
+            return $this->jsonResponse(
+                ['errors' => [_t(__CLASS__ . '.WRONG_METHOD', 'Wrong HTTP request method used')]],
+                400
```
Comment (ScopeyNZ) [inline src/Controller/AdminRegistrationController.php]:
```suggestion
                405
```

Comment (Cheddam) [reply src/Controller/AdminRegistrationController.php]: Since this can't be reached via any other HTTP method now, is it even necessary to keep this check in place?

File: src/Controller/AdminRegistrationController.php [hunk 5]
```diff
@@ -30,17 +34,28 @@ class AdminRegistrationController extends LeftAndMain
     private static $url_handlers = [
         'GET register/$Method' => 'startRegistration',
         'POST register/$Method' => 'finishRegistration',
-        'GET remove/$Method' => 'removeRegisteredMethod',
+        'DELETE remove/$Method' => 'removeRegisteredMethod',
```
Comment (Cheddam) [inline src/Controller/AdminRegistrationController.php]: `DELETE remove/:identifier` is kind of doubling up the directive - a more RESTful endpoint would be `DELETE method/:identifier`. Maybe we should adjust these if we get time.

File: src/Controller/AdminRegistrationController.php (hunk 4 above)
Comment (robbieaverill) [reply src/Controller/AdminRegistrationController.php]: Yep good point!

File: src/Controller/AdminRegistrationController.php (hunk 5 above)
Comment (robbieaverill) [reply src/Controller/AdminRegistrationController.php]: Good idea. I don't think the registration handlers should change, but this one should for sure. I don't think there's an obvious restful way to combine "set default method" with `[httpMethod] method/$Method` either, so the result would probably be `DELETE method/$Method` only for now?

Comment (Cheddam) [reply src/Controller/AdminRegistrationController.php]: In RESTful terms, setting the default method could potentially come in the form of an update to the method itself (`POST method/$Method`, with something like `{default: true}` in the body). You're right in that this would require significant changes to how that endpoint is currently implemented, so alternatively you could use `PUT` to designate an update over a creation and split the logic that way.

Another RESTish option would be to treat 'default' as a 'sub-resource' of the method - so `PUT method/$Method/default`.

Comment (robbieaverill) [reply src/Controller/AdminRegistrationController.php]: I like that last option

Comment (robbieaverill) [reply src/Controller/AdminRegistrationController.php]: Although semantically you’re updating the member not the method, since we store the default flag on the member

Comment (robbieaverill) [reply src/Controller/AdminRegistrationController.php]: But you said sub-resource, so 👍

**Comment by robbieaverill** [comment]
Patch coverage is a little down but I'm reasonably happy with the test coverage it has TBH. Happy to write more if requested though.
